### ISSUE TITLE
`ConstrainedLogitsProcessor` constraint fails when `prefix_allowed_tokens_fn` returns empty list, causing unconstrained random sampling

### ISSUE THREAD

**Issue by lyzltysgithub** [AkaliKong/MiniOneRec#50]
### Description

When using `ConstrainedLogitsProcessor` for constrained decoding (e.g., generating only valid semantic IDs), the constraint silently fails if the model generates an unexpected token. Once this happens, all subsequent tokens become completely unconstrained, resulting in random natural language output instead of the expected structured format（SID）.

### Environment

- Transformers version: 4.x
- PyTorch version: 2.x
- Python version: 3.10+

### Problem Analysis

The issue is in the `__call__` method of `ConstrainedLogitsProcessor`:

```python
def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> torch.FloatTensor:
    scores = torch.nn.functional.log_softmax(scores, dim=-1)
    mask = torch.full_like(scores, -1000000)  # Initialize: block all tokens
    
    for batch_id, beam_sent in enumerate(input_ids.view(-1, self._num_beams, input_ids.shape[-1])):
        for beam_id, sent in enumerate(beam_sent):
            hash_key = sent[-self.count:].tolist()
            prefix_allowed_tokens = self._prefix_allowed_tokens_fn(batch_id, hash_key)
            
            if len(prefix_allowed_tokens) == 0:
                continue  # ⚠️ BUG: mask remains all -1000000
            
            mask[batch_id * self._num_beams + beam_id, prefix_allowed_tokens] = 0
    
    scores = scores + mask
    return scores
```

**Two issues:**

1. **Mask value `-1000000` is not strict enough**: `exp(-1000000)` is extremely small but not exactly zero. With high temperature or numerical edge cases, non-allowed tokens can still be sampled.

2. **`continue` statement causes constraint to completely fail**: When `prefix_allowed_tokens_fn` returns an empty list (because the current sequence doesn't match any valid path), the mask remains all `-1000000`. After softmax, all tokens have approximately equal probability, causing **completely random sampling**.

### Steps to Reproduce

```python
# Example hash_dict (valid SID paths)
hash_dict = {
    "50256-35": [100, 101],        # After "Response:\n", allow <a_0> or <a_1>
    "50256-35-100": [200],         # After <a_0>, allow <b_0>
    "50256-35-100-200": [300],     # After <a_0><b_0>, allow <c_0>
}

# Step 1: Model generates first token
# hash_key = [50256, 35] → found → allowed_tokens = [100, 101]
# Due to sampling randomness, model picks token 999 ("Industry") ← shouldn't happen but can

# Step 2: Constraint fails
# hash_key = [999] → NOT found → returns []
# continue executes → mask stays all -1000000
# softmax(-1000000, -1000000, ...) → uniform distribution
# Random token selected

# Step 3+: Continues to fail
# Each subsequent step also returns [] → completely random output
# Final output: "Industry and Scientific products..." instead of "<a_0><b_0><c_0>"
```

### Expected Behavior

When the constraint cannot find valid continuations, the decoder should either:
1. Force generation of EOS token to end the sequence
2. Completely prevent invalid tokens from being selected in the first place

### Suggested Fixes

**Option 1: Use `-inf` instead of `-1000000`**
```python
mask = torch.full_like(scores, float('-inf'))  # exp(-inf) = 0, absolutely blocked
```

**Option 2: Force EOS when no valid tokens available**
```python
if len(prefix_allowed_tokens) == 0:
    mask[batch_id * self._num_beams + beam_id, eos_token_id] = 0  # Force EOS
    continue
```

**Option 3: Raise error or warning for debugging**
```python
if len(prefix_allowed_tokens) == 0:
    warnings.warn(f"No valid tokens for hash_key {hash_key}, forcing EOS")
    mask[batch_id * self._num_beams + beam_id, eos_token_id] = 0
```

### Impact

This bug causes:
- Unpredictable model outputs when using constrained decoding
- Difficult-to-debug failures (constraint silently fails)
- Wasted compute on invalid generation paths

### Related

This affects any use case requiring strict constrained generation, such as:
- Semantic ID generation for recommendation systems
- Structured output generation (JSON, code, etc.)
- Grammar-constrained decoding
